### PR TITLE
fix: fail on invalid config

### PR DESCRIPTION
**Pull request by hoffa**
*Issue #, if available:* #2311 

### Why is this change necessary?

- Prevent issues in the config from being ignored.
- Improve UX by stating the root cause as early as possible.

### How does it address the issue?

Fail immediately if a non-`KeyError` exception is thrown when processing the config.

For example, with the following `samconfig.toml`:

```toml
version = 0.1
[default.deploy.parameters]
stack_name = _sam-app"
```

Instead of cryptic error messages as in #2311, `sam deploy` and `sam deploy --guided` will now show the following output:

```bash
Error: Error reading configuration: Unexpected character: '_' at line 3 col 13
```

A few notes regarding the PR:
- Keeping the `KeyError` handling as-is for now as it's not directly relevant to the issue and there might be marginal benefit to it (e.g. allow specifying a custom `config_env` without the corresponding TOML table).
- Removed the [`sanity_check()`](https://github.com/aws/aws-sam-cli/blob/3edb55cd8dba9898ddf2e1c8bd9eeba0434c08de/samcli/lib/config/samconfig.py#L134)  as the `get_all()` below it already validates the config by calling `_read()` (which is also what `sanity_check()` does, except by swallowing exceptions).

### What side effects does this change have?

`sam deploy` fails early if there's a problem reading the config.

### Did you change a dependency in `requirements/base.txt`?

No.

### Checklist

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/cli/cli_config_file.py [hunk 1]
```diff
@@ -76,7 +75,6 @@ def __call__(self, config_path, config_env, cmd_names):
 
             # NOTE(TheSriram): change from tomlkit table type to normal dictionary,
             # so that click defaults work out of the box.
-            samconfig.sanity_check()
```
Comment (hoffa) [inline samcli/cli/cli_config_file.py:79]: Note: unnecessary as the `get_all()` below already validates the config by calling `_read()` (which is also what [`sanity_check()`](https://github.com/aws/aws-sam-cli/blob/3edb55cd8dba9898ddf2e1c8bd9eeba0434c08de/samcli/lib/config/samconfig.py#L134) does, except by swallowing the useful `tomlkit.exceptions.TOMLKitError`).

Comment (c2tarun) [inline samcli/cli/cli_config_file.py:79]: What was this doing here? We don't need it anymore?

Comment (hoffa) [reply samcli/cli/cli_config_file.py:79]: Right, I explain the why in the summary. Basically `get_all()` already does what `sanity_check()` does, except that `sanity_check()` swallows the important TOML exception with the useful error message.

**Review by c2tarun — APPROVED** [review]

**Review by jfuss — CHANGES_REQUESTED** [review]
Can we add an integ tests that covers the invalid samconfig case?

**Comment by hoffa** [comment]
Added unit/integ tests!

**Review by jfuss — APPROVED** [review]

**Review by sriram-mv — APPROVED** [review]
Nice fix!

**Pull request merged**